### ISSUE TITLE
Using ethers v6 providers results in error

### ISSUE THREAD

**Issue by hmmdeif** [Uniswap/smart-order-router#383]
- **I'm submitting a ...**
  [x] bug report
  [ ] feature request
  [ ] question about the decisions made in the repository
  [ ] question about how to use this project

- **Summary**

Latest version of ethers (6.7.1) and trying to use a provider results in `Error: invalid signer or provider`. Should be easily reproducible.

This is the code I have:

```
const router = new AlphaRouter({
    chainId: 1,
    provider: (new JsonRpcProvider('http://an.actual.rpc.url').provider) as any,
})
```

I've tried multiple providers and variations, all error.

- **Other information** (e.g. detailed explanation, stack traces, related issues, suggestions how to fix, links for us to have context, eg. StackOverflow, personal fork, etc.)

Currently a few SO recent unanswered questions:

* https://stackoverflow.com/questions/77002340/invalid-signer-or-provider-for-alpharouter
* https://ethereum.stackexchange.com/questions/153167/error-invalid-signer-or-provider-argument-signerorprovider-value-code-i

**Comment by burgossrodrigo** [comment]
Same issue here!

**Comment by ashaller2017** [comment]
same

**Comment by burgossrodrigo** [comment]
> same

What i did: insteed of using JsonRpcProvider i went to AlchemyProvider and it worked. Try it or another one like infura.

**Comment by HareemSaad** [comment]
@burgossrodrigo can you show me how you built it because this workaround did not work for me

**Comment by burgossrodrigo** [comment]
```
import { AlchemyProvider } from '@ethersproject/providers';

const new AlchemyProvider(ChainId.POLYGON_MUMBAI, YOUR_API_KEY)


````

**Comment by HareemSaad** [comment]
> ```
> import { AlchemyProvider } from '@ethersproject/providers';
> 
> const new AlchemyProvider(ChainId.POLYGON_MUMBAI, YOUR_API_KEY)
> ```
@burgossrodrigo  unfortunately, this does not work for me.

**Comment by burgossrodrigo** [comment]
> > ```
> > import { AlchemyProvider } from '@ethersproject/providers';
> > 
> > const new AlchemyProvider(ChainId.POLYGON_MUMBAI, YOUR_API_KEY)
> > ```
> 
> @burgossrodrigo unfortunately, this does not work for me.


Which error

**Comment by HareemSaad** [comment]
> > > ```
> > > import { AlchemyProvider } from '@ethersproject/providers';
> > > 
> > > const new AlchemyProvider(ChainId.POLYGON_MUMBAI, YOUR_API_KEY)
> > > ```
> > 
> > 
> > @burgossrodrigo unfortunately, this does not work for me.
> 
> Which error

`Error: invalid signer or provider` this error

**Comment by Mantas-M** [comment]
I had this issue and could not revert to using ethers@5 for my project since it was not compatible with the latest hardhat version. 

What solved the issue was using v6 for the entire project and only using the provider from v5, besides having 2 ethers versions installed, seems like a decent workaround.

I installed v5 like so
```
yarn add ethersv5@npm:ethers@^5.7.2                                                                    
```

And imported like this
```
import { ethers } from 'ethersv5'
```

Then you can setup the AlphaRouter with the v5 provider and use v6 for everything else.

**Comment by HareemSaad** [comment]
> I had this issue and could not revert to using ethers@5 for my project since it was not compatible with the latest hardhat version.
> 
> What solved the issue was using v6 for the entire project and only using the provider from v5, besides having 2 ethers versions installed, seems like a decent workaround.
> 
> I installed v5 like so
> 
> ```
> yarn add ethersv5@npm:ethers@^5.7.2                                                                    
> ```
> 
> And imported like this
> 
> ```
> import { ethers } from 'ethersv5'
> ```
> 
> Then you can setup the AlphaRouter with the v5 provider and use v6 for everything else.

Thankyou!!

**Comment by itxtoledo** [comment]
This package only works with ethers v5, see package.json

**Comment by clemsos** [comment]
Support for v6 is needed !

**Comment by jfbloom22** [comment]
I tried @Mantas-M's solution

> I had this issue and could not revert to using ethers@5 for my project since it was not compatible with the latest hardhat version.
> 
> What solved the issue was using v6 for the entire project and only using the provider from v5, besides having 2 ethers versions installed, seems like a decent workaround.
> 
> I installed v5 like so
> 
> ```
> yarn add ethersv5@npm:ethers@^5.7.2                                                                    
> ```
> 
> And imported like this
> 
> ```
> import { ethers } from 'ethersv5'
> ```
> 
> Then you can setup the AlphaRouter with the v5 provider and use v6 for everything else.

But I ran into an error: 
   `Cannot find module 'ethers/lib/utils' from '../node_modules/.pnpm/@uniswap+v4-sdk@1.0.0_hardhat@2.22.6/node_modules/@uniswap/v4-sdk/dist/v4-sdk.cjs.development.js'`

In other words, the v5 version of ethers is a peer dependency.  
Solution is to switch it around so that v5 is imported as "ethers" and v6 is imported as "ethers-v6":

```
    "ethers": "^5.7.2",
    "ethers-v6": "npm:ethers@^6.13.2",
```

**Comment by burgossrodrigo** [comment]
> I tried @Mantas-M's solution
> 
> > I had this issue and could not revert to using ethers@5 for my project since it was not compatible with the latest hardhat version.
> > What solved the issue was using v6 for the entire project and only using the provider from v5, besides having 2 ethers versions installed, seems like a decent workaround.
> > I installed v5 like so
> > ```
> > yarn add ethersv5@npm:ethers@^5.7.2                                                                    
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And imported like this
> > ```
> > import { ethers } from 'ethersv5'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then you can setup the AlphaRouter with the v5 provider and use v6 for everything else.
> 
> But I ran into an error: `Cannot find module 'ethers/lib/utils' from '../node_modules/.pnpm/@uniswap+v4-sdk@1.0.0_hardhat@2.22.6/node_modules/@uniswap/v4-sdk/dist/v4-sdk.cjs.development.js'`
> 
> In other words, the v5 version of ethers is a peer dependency. Solution is to switch it around so that v5 is imported as "ethers" and v6 is imported as "ethers-v6":
> 
> ```
>     "ethers": "^5.7.2",
>     "ethers-v6": "npm:ethers@^6.13.2",
> ```

Impressive.

**Comment by ohnodev** [comment]
Yes this worked for me. I did it backwards though. So I use npm and my package.json looks like this:

"ethers": "^6.13.2",
 "ethersv5": "npm:ethers@^5.7.2",

So just import ethers into the rest of your project, or it's probably already imported that way, and add the ethersv5 library only to the file you want the v5 provider for. If you want to abstract it more can write a separate file that only imports ethersv5 and exports the provider, then you can use it within your project where you need.
